fauna/client/client.py

```diff
@@ -579,6 +609,50 @@ def _set_endpoint(self, endpoint):
     self._endpoint = endpoint
 
 
+class StreamIterator:
+  """A class that mix a ContextManager and an Iterator so we can detected retryable errors."""
+
+  def __init__(self, client, fql):
```

Comment:
The contract and impl in this class feels a little awkward. The main issue I see is that we accept a client, then call a private method on the client as part of the impl.

I think we can accept headers, the _http client_ instead of the fauna client, and the fql. Then we've got a clean separation. You can get rid of client._stream entirely and implement it here, and you don't need to set the headers every time (you can do that once in client.stream).

Let me know what you think. Maybe you had plans that would be ruined by this, and if so, we can talk through other approaches.